src/enclave/enclave_oe.c

```diff
@@ -311,38 +285,87 @@ static void _read_eeid_config()
     sgxlkl_enclave_state.config = cfg;
 }
 
+void sgxlkl_ensure_inside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_within_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is not in enclave memory.\n");
+}
+
+void sgxlkl_ensure_outside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_outside_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is in enclave memory.\n");
+}
+
```

Comment:
Regarding the placement of `oe_lfence()`: One option is to place it in `sgxlkl_ensure_outside()` which would ensure its use for all outside memory checks.  The other option is to place `oe_lfence()` at select locations that aggregate multiple mem checks. The former is preferred for completeness and simplicity, while the latter may improve performance slightly.
For recent discussion of the need for manually inserted lfences, please see:
https://github.com/openenclave/openenclave/issues/3401
and the discussion:
https://github.com/openenclave/openenclave/pull/3418

Reply:
I added oe_lfence()s to the SGX-LKL memory boundary checks for good measure. We can remove or adjust them once https://github.com/openenclave/openenclave/pull/3418 comes to fruition.